test/e2e/persistent_volumes.go

```diff
@@ -66,58 +64,111 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
 			persistentVolumeTestCleanup(c, config)
 		}()
 
+		// Create the nfs server pod
 		pod := startVolumeServer(c, config)
 		serverIP := pod.Status.PodIP
 		framework.Logf("NFS server IP address: %v", serverIP)
 
+		// Define the PersistentVolume and PersistentVolumeClaim
 		pv := makePersistentVolume(serverIP)
 		pvc := makePersistentVolumeClaim(ns)
 
-		framework.Logf("Creating PersistentVolume using NFS")
+		// Create the PersistentVolume and wait for PersistentVolume.Status.Phase to be Available
+		framework.Logf("Creating PersistentVolume")
 		pv, err := c.PersistentVolumes().Create(pv)
-		Expect(err).NotTo(HaveOccurred())
+		if err != nil {
+			framework.Failf("Create PersistentVolume failed: %v", err)
+		}
+		framework.WaitForPersistentVolumePhase(api.VolumeAvailable, c, pv.Name, 1*time.Second, 20*time.Second)
 
+		// Create the PersistentVolumeClaim
 		framework.Logf("Creating PersistentVolumeClaim")
 		pvc, err = c.PersistentVolumeClaims(ns).Create(pvc)
-		Expect(err).NotTo(HaveOccurred())
-
-		// allow the binder a chance to catch up.  should not be more than 20s.
-		framework.WaitForPersistentVolumePhase(api.VolumeBound, c, pv.Name, 1*time.Second, 30*time.Second)
+		if err != nil {
+			framework.Failf("Create PersistentVolumeClaim failed: %v", err)
+		}
 
+		// Wait for PersistentVolume.Status.Phase to be Bound. Can take several minutes.
+		err = framework.WaitForPersistentVolumePhase(api.VolumeBound, c, pv.Name, 3*time.Second, 300*time.Second)
+		if err != nil {
+			framework.Failf("PersistentVolume failed to enter a bound state: %+v", err)
+		}
+		// Check the PersistentVolume.ClaimRef.UID for non-nil value as confirmation of the bound state.
+		framework.Logf("Checking PersistentVolume ClaimRef is non-nil")
 		pv, err = c.PersistentVolumes().Get(pv.Name)
-		Expect(err).NotTo(HaveOccurred())
-		if pv.Spec.ClaimRef == nil {
-			framework.Failf("Expected PersistentVolume to be bound, but got nil ClaimRef: %+v", pv)
+		if pv.Spec.ClaimRef == nil || len(pv.Spec.ClaimRef.UID) == 0 {
+			pvJson, _ := json.MarshalIndent(pv, "", "  ")
+			framework.Failf("Expected PersistentVolume to be bound, but got nil ClaimRef or UID: %+v", string(pvJson))
+		}
+
+		// Check the PersistentVolumeClaim.Status.Phase for Bound state
+		framework.Logf("Checking PersistentVolumeClaim status is Bound")
+		pvc, err = c.PersistentVolumeClaims(ns).Get(pvc.Name)
```

Comment:
During binding the controller saves bound PV first and bound PVC after that. It's not the same etcd transaction and it could happen that the test gets to CPU when PV is saved and `WaitForPersistentVolumePhase` few lines above returns, but PVC is not saved yet.

Please add `WaitForPersistentVolumeClaimPhase` here to make sure the controller had time to save both PVC and PV.